File: pkg/app/server/network/handlers/firewall.go
/******************************************************************************
 * Copyright (c) KylinSoft Co., Ltd.2021-2022. All rights reserved.
 * PilotGo is licensed under the Mulan PSL v2.
 * You can use this software accodring to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *     http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN 'AS IS' BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * Author: zhanghan
 * Date: 2022-02-17 02:43:29
 * LastEditTime: 2022-04-13 01:51:51
 * Description: provide agent firewall manager functions.
 ******************************************************************************/
package handlers

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"openeluer.org/PilotGo/PilotGo/pkg/app/server/agentmanager"
	"openeluer.org/PilotGo/PilotGo/pkg/common/response"
)

func FirewalldConfig(c *gin.Context) {
	uuid := c.Query("uuid")

	agent := agentmanager.GetAgent(uuid)
	if agent == nil {
		response.Response(c, http.StatusOK, 400, nil, "获取uuid失败!")
		return
	}

	config, Err, err := agent.FirewalldConfig()
	if len(Err) != 0 || err != nil {
		response.Response(c, http.StatusOK, 400, gin.H{"error": Err}, "获取防火墙配置失败!")
		return
	}
	response.Response(c, http.StatusOK, 200, gin.H{"firewalld_config": config}, "获取防火墙配置成功!")
}

func FirewalldRestart(c *gin.Context) {
	uuid := c.Query("uuid")

	agent := agentmanager.GetAgent(uuid)
	if agent == nil {
		response.Response(c, http.StatusOK, 400, nil, "获取uuid失败!")
		return
	}

	restart, Err, err := agent.FirewalldRestart()
	if len(Err) != 0 || err != nil {
		response.Response(c, http.StatusOK, 400, gin.H{"error": Err}, "重启防火墙失败")
		return
	}
	response.Response(c, http.StatusOK, 200, gin.H{"firewalld_restart": restart}, "重启防火墙成功!")
}

func FirewalldStop(c *gin.Context) {
	uuid := c.Query("uuid")

	agent := agentmanager.GetAgent(uuid)
	if agent == nil {
		response.Response(c, http.StatusOK, 400, nil, "获取uuid失败!")
		return
	}

	stop, Err, err := agent.FirewalldStop()
	if len(Err) != 0 || err != nil {
		response.Response(c, http.StatusOK, 400, gin.H{"error": Err}, "关闭防火墙失败!")
		return
	}
	response.Response(c, http.StatusOK, 200, gin.H{"firewalld_stop": stop}, "关闭防火墙成功!")
}

type ZonePort struct {
	UUID string `json:"uuid"`
	Zone string `json:"zone"`
	Port string `json:"port"`
}

func FirewalldZonePortAdd(c *gin.Context) {
	var zp ZonePort
	c.ShouldBind(&zp)
	agent := agentmanager.GetAgent(zp.UUID)
	if agent == nil {
		response.Response(c, http.StatusOK, 400, nil, "获取uuid失败!")
		return
	}

	add, Err, err := agent.FirewalldZonePortAdd(zp.Zone, zp.Port)
	if len(Err) != 0 || err != nil {
		response.Response(c, http.StatusOK, 400, gin.H{"error": Err}, "添加失败!")
		return
	}
	response.Response(c, http.StatusOK, 200, gin.H{"firewalld_add": add}, "添加成功!")
}

func FirewalldZonePortDel(c *gin.Context) {
	var zp ZonePort
	c.ShouldBind(&zp)

	agent := agentmanager.GetAgent(zp.UUID)
	if agent == nil {
		response.Response(c, http.StatusOK, 400, nil, "获取uuid失败!")
		return
	}

	del, Err, err := agent.FirewalldZonePortDel(zp.Zone, zp.Port)
	if len(Err) != 0 || err != nil {
		response.Response(c, http.StatusOK, 400, gin.H{"error": Err}, "删除失败!")
		return
	}
	response.Response(c, http.StatusOK, 200, gin.H{"firewalld_del": del}, "删除成功!")
}

File: frontend/vue/request/firewall.js
/*
 * Copyright (c) KylinSoft Co., Ltd.2021-2022. All rights reserved.
 * PilotGo is licensed under the Mulan PSL v2.
 * You can use this software accodring to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *     http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN 'AS IS' BASIS, WITHOUT WARRANTIES OF ANY KIND, 
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * @Author: zhaozhenfang
 * @Date: 2022-04-18 15:18:50
 * @LastEditTime: 2022-07-01 08:51:00
 */
import request from './request'
// 重启防火墙
export function reStart(data) {
  return request({
    url: 'agent/firewall_restart',
    method: 'get',
    params: data
  })
}

// 关闭防火墙
export function close(data) {
  return request({
    url: 'agent/firewall_stop',
    method: 'get',
    params: data
  })
}

// 指定区域开放端口
export function openPort(data) {
  return request({
    url: 'agent/firewall_addzp',
    method: 'post',
    data
  })
}


// 删除开放的端口
export function deleteOpenPort(data) {
  return request({
    url: 'agent/firewall_delzp',
    method: 'post',
    data
  })
}

// 获取防火墙配置
export function getConfig(data) {
  return request({
    url: 'api/firewall_config',
    method: 'get',
    params: data
  })
}

// 更改防火墙默认区域
export function changeDefaultZone(data) {
  return request({
    url: 'agent/firewall_default',
    method: 'post',
    data
  })
}

// 获取防火墙区域配置
export function getZoneConfig(data) {
  return request({
    url: 'api/firewall_zone',
    method: 'get',
    params: data
  })
}

// 添加服务
export function addService(data) {
  return request({
    url: 'agent/firewall_serviceAdd',
    method: 'post',
    data
  })
}

// 删除服务
export function delService(data) {
  return request({
    url: 'agent/firewall_serviceRemove',
    method: 'post',
    data
  })
}
// 添加来源
export function addSource(data) {
  return request({
    url: 'agent/firewall_sourceAdd',
    method: 'post',
    data
  })
}

// 删除来源
export function delSource(data) {
  return request({
    url: 'agent/firewall_sourceRemove',
    method: 'post',
    data
  })
}